docs/protocol/src/crypto/tct.md

```diff
@@ -0,0 +1,383 @@
+\end{align}
+$$
+
+where the function $\mathtt{chunk}$ is defined on non-empty sequences as:
+
+$$
+\begin{align}
+\mathtt{chunk}([a, b, c, d, \dots rest]) & = [\mathtt{Node}(a, b, c, d), \dots \mathtt{chunk}(rest)] \\
+\mathtt{chunk}([a, b, c]) & = [\mathtt{Node}(a, b, c)] \\
+\mathtt{chunk}([a, b]) & = [\mathtt{Node}(a, b)] \\
+\mathtt{chunk}([a]) & = [\mathtt{Node}(a)] \\
+\end{align}
+$$
+
+In English, this algorithm is:
+
+- if the sequence is empty, return the empty tier;
+- otherwise, do the following exactly 8 times:
+  - in left to right order, greedily partition the sequence into chunks of at most 4 items, so
+    that all but the last chunk has exactly 4 items, and the last chunk has at most 4 items;
+  - for each chunk, construct a single node from it; and
+  - unless done with the 8 iterations, repeat again, starting with this sequence of nodes (which is
+    now shorter by about a factor of 4)
+- if the input sequence was less than $4^8$ items long, the resultant sequence will now contain
+  exactly one node; return it as the top-most node contained in a tier.
+
+Constructing the entire tree requires mapping the $\mathtt{tier}$ function over the triply-nested
+sequence:
+
+$$
+\begin{align}
+& \mathtt{tree}(eternity) =\\
+& \ \ \ \ \mathtt{tier}\left(eternity\mathtt{.map}(epoch \Rightarrow \mathtt{tier}(epoch\mathtt{.map}(block \Rightarrow
+\mathtt{tier}(block)))))\right)
+\end{align}
+$$
+
+where the $\mathtt{map}$ function applies a function to each element of a sequence.
+
+## Hashing Nodes
+
+Each commitment is represented by a single element from the $\mathbb{F}_q$ field of
+[BLS12-377](https://docs.rs/ark-ed-on-bls12-377/0.3.0/ark_ed_on_bls12_377/). The [Poseidon for
+BLS12-377](./poseidon.md) hash function is defined over that same field.
+
+### Elementary Hashes and Domain Separators
+
+We define the hash of each node in the tree in terms of two arities of the Poseidon hash function:
+
+- $\mathtt{hash_1} : \mathbb{F}_q \times \mathbb{F}_q \to \mathbb{F}_q$
+- $\mathtt{hash_4} : \mathbb{F}_q \times (\mathbb{F}_q \times \mathbb{F}_q \times \mathbb{F}_q
+  \times \mathbb{F}_q) \to \mathbb{F}_q$
+
+Each of these takes as its first argument a _domain separator_. For the tiered note commitment tree,
+we use a domain separator calculated in terms of the _base domain separator_ $\mathbf{D}_0 =
+\mathtt{blake2b}(\mathtt{"penumbra.tct"})$ interpreted in little-endian order as an element of
+$\mathbb{F}_q$, i.e. the following hash output from
+[BLAKE2](https://datatracker.ietf.org/doc/html/rfc7693):
+
+```plain_text
+0x197e501077d94b004b29e91196b9da397b8acbc98ac2d13d9099ad99f5215e490f171a0406f223ed25bafe4bc0451b649ba4b3fe0374d4a0b1ffd2ead37f531e
+```
+
+...interpreted as a (large) integer in little-endian order, and reduced by the modulus of
+$\mathbb{F}_q$.
```

Comment:
We can skip the test vectors here, since they're not enough to determine the implementation behavior anyways, and so it mostly serves to clutter the description.